src/ir/struct-utils.h

```diff
@@ -0,0 +1,251 @@
+//   in another object).
+//
+//   void noteCopy(HeapType type, Index index, T& info);
+//
+// We track information from struct.new and struct.set separately, because in
+// struct.new we know more about the type - we know the actual exact type being
+// written to, and not just that it is of a subtype of the instruction's type,
+// which helps later.
+template<typename T, typename SubType>
+struct Scanner : public WalkerPass<PostWalker<Scanner<T, SubType>>> {
+  bool isFunctionParallel() override { return true; }
+
+  Scanner(FunctionStructValuesMap<T>& functionNewInfos,
+          FunctionStructValuesMap<T>& functionSetInfos)
+    : functionNewInfos(functionNewInfos), functionSetInfos(functionSetInfos) {}
+
+  void visitStructNew(StructNew* curr) {
+    auto type = curr->type;
+    if (type == Type::unreachable) {
+      return;
+    }
+
+    // Note writes to all the fields of the struct.
+    auto heapType = type.getHeapType();
+    auto& fields = heapType.getStruct().fields;
+    auto& infos = functionNewInfos[this->getFunction()][heapType];
+    for (Index i = 0; i < fields.size(); i++) {
+      if (curr->isWithDefault()) {
+        static_cast<SubType*>(this)->noteDefault(
+          fields[i].type, heapType, i, infos[i]);
+      } else {
+        noteExpressionOrCopy(curr->operands[i], heapType, i, infos[i]);
+      }
+    }
+  }
+
+  void visitStructSet(StructSet* curr) {
+    auto type = curr->ref->type;
+    if (type == Type::unreachable) {
+      return;
+    }
+
+    // Note a write to this field of the struct.
+    noteExpressionOrCopy(
+      curr->value,
+      type.getHeapType(),
+      curr->index,
+      functionSetInfos[this->getFunction()][type.getHeapType()][curr->index]);
+  }
+
+  void
+  noteExpressionOrCopy(Expression* expr, HeapType type, Index index, T& info) {
+    // Look at the value falling through, if it has the exact same type
+    // (otherwise, we'd need to consider both the type actually written and the
+    // type of the fallthrough, somehow).
+    auto* fallthrough = Properties::getFallthrough(
+      expr, this->getPassOptions(), *this->getModule());
+    if (fallthrough->type == expr->type) {
+      expr = fallthrough;
+    }
+    if (auto* get = expr->dynCast<StructGet>()) {
+      if (get->index == index && get->ref->type != Type::unreachable &&
+          get->ref->type.getHeapType() == type) {
+        static_cast<SubType*>(this)->noteCopy(type, index, info);
+        return;
+      }
+    }
+    static_cast<SubType*>(this)->noteExpression(expr, type, index, info);
+  }
+
+  FunctionStructValuesMap<T>& functionNewInfos;
+  FunctionStructValuesMap<T>& functionSetInfos;
+};
+
+// Helper class to propagate information about fields to sub- and/or super-
+// classes. While propagating it calls a method
+//
+//  to.combine(from)
+//
+// which combines the information from |from| into |to|, and should return true
+// if we changed something.
+template<typename T> class StructValuePropagator {
```

Comment:
Can we get something about `TypeHierarchy` into the name to hint at how the propagation is happening?

Reply:
Ah, nice idea, renaming to `TypeHierarchyPropagator`.